114 get users location lobby

tamari-gray opened the pull request:
fixes #114 

- getting users location to use for distance calculations in #60

Commit:
get users location in lobby before render gamelist

Commit:
started testing

Commit:
mockservice for testing

Commit:
test get users location + extract listofgames

Commit:
minor test fix

Commit:
Merge branch 'dev' into 114-get-users-location-lobby

Commit:
added lobby route (think i deleted when merging)

tamari-gray commented on lib/screens/lobby/lobby.dart:
```diff
@@ -18,9 +22,9 @@ class LobbyScreen extends StatelessWidget {
+              // create a function to call and sign out on confirmation
               final signOut = () async {
-                context.read<Navigation>().pop();
+                Navigator.of(context).pop();
```
oops, gotta change that back

tamari-gray commented on lib/screens/lobby/lobby.dart:
```diff
@@ -34,27 +38,22 @@ class LobbyScreen extends StatelessWidget {
+            if (snapshot.hasData == false) {
+              return Loading();
+            } else {
+              final userLocation = Location(
```
location will be passed to list of created games to do distance calculations in #60

nickmeinhold replied:
sweet!

tamari-gray commented on lib/services/location_service.dart:
```diff
@@ -11,7 +11,7 @@ class LocationService {
 
   // get users currentLocation
-  Future<Position> getUsersCurrentLocation() =>
+  Future<Position> get getUsersCurrentLocation =>
```
just a wee update

nickmeinhold replied:
I think standard practice for a getter is to not have 'get' in the name.  I believe getters are conceptually the same as accessing a member and so should just be the member name.

tamari-gray replied:
sweet thank you!

tamari-gray commented on test/create_game_section/Lobby/lobby_widget_test.dart:
```diff
@@ -7,16 +7,64 @@ import 'package:flutter_test/flutter_test.dart';
+  });
   testWidgets(
       'shows loading icon + redirects to welcome screen on successfull logout',
       (WidgetTester tester) async {
```
all added code under this comment was to fix existing test.
(sorry if im just pointing out the obvious)

Commit:
replaced navigation with navigation service method

tamari-gray commented on lib/screens/lobby/lobby.dart:
```diff
@@ -34,27 +38,22 @@ class LobbyScreen extends StatelessWidget {
           )
         ],
       ),
-      body: Container(
```
extracted red block below into its own file

nickmeinhold replied:
nice!

tamari-gray reviewed (commented):
woohoo shes ready for a review

nickmeinhold commented on lib/screens/lobby/list_of_created_games.dart:
```diff
@@ -0,0 +1,32 @@
+          // show empty screen if stream hasnt come through yet
+          if (snapshot.data == null) {
+            return Container(
+              key: Key('list_of_created_games_empty_screen'),
```
Nice!  I never thought to give the empty state a key so you can check for it... love it!

nickmeinhold commented on lib/screens/lobby/list_of_created_games.dart:
```diff
@@ -0,0 +1,32 @@
+          }
+        });
+  }
+}
```
I liked this whole file - nice clean code I thought 👍

tamari-gray replied:
thank you :pray: 🙏

nickmeinhold commented on lib/screens/lobby/lobby.dart:
```diff
@@ -1,12 +1,16 @@
-import 'package:niira/models/game.dart';
-import 'package:niira/navigation/navigation.dart';
-import 'package:niira/screens/lobby/created_game_tile.dart';
+import 'package:geolocator/geolocator.dart';
```
I reckon you're be better off removing the geolocator dependency - it's a clean architecture thing, happy to have a crack at explaining the benefits if you want, not sure if I'll do a good job or not. I can point you to a good reference if that's more helpful. Uncle Bob's books would be the best source I'd say, it's mainly the 'D' in the SOLID principles that's relevant here I'd say, off the top of my head... "Depend on abstractions, not on concretions."  [SOLID Principles made easy | Hacker Noon](https://hackernoon.com/solid-principles-made-easy-67b1246bcdf)

nickmeinhold replied:
So you'd have to make the LocationService convert (the geolocator) Position to (your homemade) Location objects so your return types can use Location (or something else that is your own PODO)

nickmeinhold replied:
But it's your call of course!

tamari-gray replied:
thanks for pointing it out! Super happy to implement that 😄

nickmeinhold commented on test/mocks/services/mock_location_service.dart:
```diff
@@ -0,0 +1,20 @@
+import 'package:niira/models/location.dart';
+import 'package:niira/services/location_service.dart';
+
+class MockLocationService extends Mock implements LocationService {
```
If you are overriding the methods of the class that suggests to me that the mock might be better as a fake or a real object. What do you think?

tamari-gray replied:
i like it, just refreshed my brain with mockito docs and im super keen to change it to a fake :)

nickmeinhold reviewed (changes requested):
Hey mate!  Awesome work!  Had a couple of suggestions and questions but let me know if you'd prefer to keep moving and I'll hit the Approve button 👍

Commit:
implemented suggestions from review

tamari-gray commented:
thanks for the SOLID review @nickmeinhold!
i added some changes and cleaned up her architecture 👷 🏗️ (think uncle bob will be proud of us cutting out that Position object)
shes all ready for another look (minor changes), when you are. thank you!

nickmeinhold reviewed (approved):
Bloody good work mate... and so fast 💨

nickmeinhold commented:
oh and just a quick appreciation 👏 for the comedy... absolutely lovin the wordplay!
